### ISSUE TITLE
Strong name validation failed on FSharp.Core 4.4.3.0

### ISSUE THREAD

**Issue by cartermp** [Microsoft/visualfsharp#4263]
```
System.IO.FileLoadException: Could not load file or assembly 'FSharp.Core, Version=4.4.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. Strong name validation failed. 
```

https://www.nuget.org/packages/FSharp.Core/

What's going on here?

**Comment by btrepp** [comment]
```
System.IO.FileLoadException: Could not load file or assembly 'FSharp.Core, Version=4.4.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. Strong name validation failed. (Exception from HRESULT: 0x8013141A) ---> System.Security.SecurityException: Strong name validation failed. (Exception from HRESULT: 0x8013141A)
```

**Comment by forki** [comment]
yes same here. reverting paket upgrade.

**Comment by dsyme** [comment]
After being notified about this and after reviewing the above I decided unlisted the package https://www.nuget.org/packages/FSharp.Core/4.3.0.

**Comment by KevinRansom** [comment]
Ouch!!!!

It turns out that we started creating the nuget packages before we sign the assemblies in them.  @brettfo and I will see what we need to do to fix that.  

This was missed because sn-Vr is still in widespread use, curses ... and ... dotnet cli doesn't really care about the signing hashes.

Sorry for the inconvenience, we will address it quickly.

Kevin

**Comment by forki** [comment]
On the bright side: this was caught quickly after release. Also it's nice
that a netstandard 2 version is coming.

Am 25.01.2018 19:51 schrieb "Kevin Ransom (msft)" <notifications@github.com

**Comment by KevinRansom** [comment]
Okay ... pushed 4.3.1 to nugget with signed binaries.

@forki, The serialization issue was a compiler bug, that is now fixed and does **not require** a netstandard 2.0 build.  Should be shipped in the next.

**Comment by forki** [comment]
But there was a change in dependencies in the nuspec that looked like you
shipped 2.0 support.

Am 25.01.2018 23:10 schrieb "Kevin Ransom (msft)" <notifications@github.com
>:

Okay ... pushed 4.3.1 to nugget with signed binaries.

@forki <https://github.com/forki>, The serialization issue was a compiler
bug, that is now fixed and does *not require* a netstandard 2.0 build.
Should be shipped in the next.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/Microsoft/visualfsharp/issues/4263#issuecomment-360617401>,
or mute the thread

**Comment by forki** [comment]
Ok where is the difference between 4.3.1 package and 4.3.2 package?

**Comment by forki** [comment]
This seems to be fixed. (At least our build is now green again)  Thank you very much!

**Comment by cartermp** [comment]
Closing as the immediate issue is fixed. Linking #4269 as the long-term solution

**Comment by cartermp** [comment]
From Preview 3

![image](https://user-images.githubusercontent.com/6309070/35535139-7a054fb4-04f7-11e8-9547-cb1344a5d442.png)

**Comment by forki** [comment]
Yes I also saw issues when I pushed to azure functions. I suggest we unlist 4.3.x

**Comment by KevinRansom** [comment]
@forki, What were the issues?

**Comment by forki** [comment]
It could not find the dll while I was pretty sure it was alongside. Paket
downgrade of fsharp.core fixed it immediately.

Am 30.01.2018 08:38 schrieb "Kevin Ransom (msft)" <notifications@github.com

**Comment by KevinRansom** [comment]
Could it not find the FSharp.Core.dll or was there a manifest error or signing error?
What was the .exe?
Was it coreclr or desktop, if desktop what was the app.config?

**Comment by forki** [comment]
It said it could not find it - without further details shown. But it was
clearly there. Regarding app.config - unfortunately azure functions does
not support redirects. And this was on their beta runtime which is dotnet
core.

Am 30.01.2018 08:43 schrieb "Kevin Ransom (msft)" <notifications@github.com

**Comment by KevinRansom** [comment]
Could you try again and capture the output.

**Comment by forki** [comment]
no I'm rather not changing it before the big presentation next week. After that I can do it - but there were no interesting details shown.

**Comment by KevinRansom** [comment]
sounds good.

**Comment by forki** [comment]
Could not load file or assembly 'FSharp.Core, Version=4.4.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621). Could not load file or assembly 'FSharp.Core, Version=4.4.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.

**Comment by forki** [comment]
even in the azure functions logs there is nothing more to see.

**Comment by KevinRansom** [comment]
@cartermp, @forki.

Can someone give me straightforward repro steps to see this error.  Or better yet an oss repro that demonstrates this error.

My intuition is that this is a publish error with the app having a dependence on the 4.4.1.0 dll, and a component dependent on the 4.4.3.0 dll.  I think I will need to look at the moving parts to see how the error is introduced.

Kevin

**Comment by forki** [comment]
No I can't. I was only able to reproduce on Microsoft azure functions. But
it did repro there every single try with many different azure functions. I
needed to revert them all.

Am 31.01.2018 20:22 schrieb "Kevin Ransom (msft)" <notifications@github.com

**Comment by cartermp** [comment]
@KevinRansom Does the repro link at the top not do this for you? The repro steps are to run ctrl-f5 on that project with 15.6 Preview 3.

**Comment by cartermp** [comment]
Argh, I'm stupid. I uploaded the wrong solution that repros this on my machine. I'll do that when I get home.

**Comment by KevinRansom** [comment]
Thanks ... and probably not stupid, just a small error.

**Comment by cartermp** [comment]
Updated here: 
[ConsoleApplication18.zip](https://github.com/Microsoft/visualfsharp/files/1686681/ConsoleApplication18.zip)

**Comment by KevinRansom** [comment]
Okay, I know what is going on here.

The solution was created with an earlier version of VS and so the console app targets FSharp.Core: 4.4.1.0
````
<snip>
    <TargetFSharpCoreVersion>4.4.1.0</TargetFSharpCoreVersion>
</snip>
````

The netsdk project targets the latest version of fsharp.core.dll, from FSharp.Core nuget package 4.3.

The AutoGenerate binding redirects, for the console application says .... Hmmm!!!! you picked FSharp.Core.dll, v 4.4.1.0 and yet your dependency needs 4.4.3.0, I'm going to pick 4.4.1.0 and warn you.  Which it does:
````
2>C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(2058,5): warning MSB3277: Found conflicts between different versions of "FSharp.Core" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed.
````

The issue arises because legacy windows projects always picks a specific version of FSharp.Core and dotnet sdk, floats using nuget.

There are two choices to solve this:
1.   Edit desktop project to have the line:
````
<snip>
    <TargetFSharpCoreVersion>4.4.3.0</TargetFSharpCoreVersion>
</snip>
````
or edit the netsdk project to have the line:
````
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <FSharpCoreImplicitPackageVersion>4.2.*</FSharpCoreImplicitPackageVersion>
  </PropertyGroup>

````
I hope this helps:

**Comment by forki** [comment]
I don't think it's what my issue is. Paket always pins versions - so nuget can't float.

**Comment by KevinRansom** [comment]
If you have a repro that I could examine, or maybe zip up the deployed files, I can examine your failure.

**Comment by forki** [comment]
I wonder how to proceed. it only happens on azure functions for me.

**Comment by dsyme** [comment]
@forki I think you might have to report a bug on Azure Functions

**Comment by forki** [comment]
Yeah. But all these reports are going to dev/null in the past. Sorry I have zero confidence in that process.

**Comment by cartermp** [comment]
@forki this is the v2 runtime, yes?

**Comment by forki** [comment]
Yes it's beta runtime.

**Comment by jakzale** [comment]
@forki If you try to load a dll that is already included in azure functions runtime, but at a different version, the runtime will prevent that dll from being loaded. (Which is a well known source of pain Azure/azure-functions-host#992)

Since the V2 runtime comes with FSharp.Core.dll version 4.4.1.0, it will prevent loading FSharp.Core.dll at version 4.4.3.0 and fail with such criptic message.

If you pin FSharp.Core to 4.1.18, then it should work.

**Comment by forki** [comment]
I'm well aware of that issue, but I didn't assume they bundle a fsharp.core. That's weird. Also I'm currently using fsharp.core nuget package 4.2.3 which works. But upgrade does not.

**Comment by cartermp** [comment]
Actually the repro project in the OP is not Azure Functions, but I think we can consider this issue out of date now that we have shipped an updated FSharp.Core binary in VS
